### PR TITLE
[AppKit] Add nullability to the manual code.

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

File: src/AppKit/NSAccessibility.cs [hunk 1]
```diff
@@ -90,13 +92,13 @@ public static void PostNotification (NSObject element, NSString notification)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityRoleDescription (IntPtr role, IntPtr subrole);
 
-		public static string GetRoleDescription (NSString role, NSString subrole)
+		public static string? GetRoleDescription (NSString? role, NSString? subrole)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:95]: According to the code, passing null for `role` isn't allowed.

File: src/AppKit/NSAccessibility.cs [hunk 2]
```diff
@@ -76,12 +78,12 @@ public static void PostNotification (NSObject element, NSString notification, NS
 		[DllImport (Constants.AppKitLibrary)]
 		static extern void NSAccessibilityPostNotification (IntPtr element, IntPtr notification);
 
-		public static void PostNotification (NSObject element, NSString notification)
+		public static void PostNotification (NSObject? element, NSString? notification)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:81]: Neither `element` nor `notification` should be null according to the code.

File: src/AppKit/NSAccessibility.cs [hunk 3]
```diff
@@ -56,16 +58,16 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern void NSAccessibilityPostNotificationWithUserInfo (IntPtr element, IntPtr notification, IntPtr userInfo);
 
-		public static void PostNotification (NSObject element, NSString notification, NSDictionary userInfo)
+		public static void PostNotification (NSObject? element, NSString? notification, NSDictionary? userInfo)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:61]: According to the code, neither `element` nor `notification` should be null.

File: src/AppKit/NSAccessibility.cs [hunk 4]
```diff
@@ -56,16 +58,16 @@ public static CGPoint GetPointInView (NSView parentView, CGPoint point)
...
 		{
-			if (element == null)
+			if (element is null)
 				throw new ArgumentNullException ("element");
 
-			if (notification == null)
+			if (notification is null)
 				throw new ArgumentNullException ("notification");
 
 			IntPtr userInfoHandle;
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:69]: This can be simplified as:

```cs
var userInfoHandle = userInfo.GetHandle ();

File: src/AppKit/NSAccessibility.cs [hunk 5]
```diff
@@ -108,9 +110,9 @@ public static string GetRoleDescription (NSString role, NSString subrole)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityRoleDescriptionForUIElement (IntPtr element);
 
-		public static string GetRoleDescription (NSObject element)
+		public static string? GetRoleDescription (NSObject? element)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:113]: `element` shouldn't be null according to the code.

File: src/AppKit/NSAccessibility.cs [hunk 6]
```diff
@@ -120,9 +122,9 @@ public static string GetRoleDescription (NSObject element)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityActionDescription (IntPtr action);
 
-		public static string GetActionDescription (NSString action)
+		public static string? GetActionDescription (NSString? action)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:125]: Same

File: src/AppKit/NSAccessibility.cs [hunk 7]
```diff
@@ -132,9 +134,9 @@ public static string GetActionDescription (NSString action)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityUnignoredAncestor (IntPtr element);
 
-		public static NSObject GetUnignoredAncestor (NSObject element)
+		public static NSObject? GetUnignoredAncestor (NSObject? element)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:137]: Same

File: src/AppKit/NSAccessibility.cs [hunk 8]
```diff
@@ -144,9 +146,9 @@ public static NSObject GetUnignoredAncestor (NSObject element)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityUnignoredDescendant (IntPtr element);
 
-		public static NSObject GetUnignoredDescendant (NSObject element)
+		public static NSObject? GetUnignoredDescendant (NSObject? element)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:149]: Same

File: src/AppKit/NSAccessibility.cs [hunk 9]
```diff
@@ -157,9 +159,9 @@ public static NSObject GetUnignoredDescendant (NSObject element)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityUnignoredChildren (IntPtr originalChildren);
 
-		public static NSObject [] GetUnignoredChildren (NSArray originalChildren)
+		public static NSObject [] GetUnignoredChildren (NSArray? originalChildren)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:162]: Same

File: src/AppKit/NSAccessibility.cs [hunk 10]
```diff
@@ -170,9 +172,9 @@ public static NSObject GetUnignoredDescendant (NSObject element)
 		[DllImport (Constants.AppKitLibrary)]
 		static extern IntPtr NSAccessibilityUnignoredChildrenForOnlyChild (IntPtr originalChild);
 
-		public static NSObject [] GetUnignoredChildren (NSObject originalChild)
+		public static NSObject [] GetUnignoredChildren (NSObject? originalChild)
```
Comment (rolfbjarne) [inline src/AppKit/NSAccessibility.cs:175]: Same

File: src/AppKit/NSAlert.cs [hunk 11]
```diff
@@ -88,13 +90,13 @@ public nint RunSheetModal (NSWindow window)
 			return RunSheetModal (window, NSApplication.SharedApplication);
 		}
 
-		public nint RunSheetModal (NSWindow window, NSApplication application)
+		public nint RunSheetModal (NSWindow? window, NSApplication? application)
```
Comment (rolfbjarne) [inline src/AppKit/NSAlert.cs:93]: `application` shouldn't be null

File: src/AppKit/NSCell.cs [hunk 12]
```diff
@@ -46,9 +48,9 @@ public partial class NSCell {
 			bool vertical, NSCompositingOperation op, nfloat alphaFraction, bool flipped)
 		{
 			NSDrawThreePartImage (
-				frame, startCap != null ? startCap.Handle : IntPtr.Zero,
-				centerFill != null ? centerFill.Handle : IntPtr.Zero,
-				endCap != null ? endCap.Handle : IntPtr.Zero,
+				frame, startCap is not null ? startCap.Handle : IntPtr.Zero,
+				centerFill is not null ? centerFill.Handle : IntPtr.Zero,
+				endCap is not null ? endCap.Handle : IntPtr.Zero,
```
Comment (rolfbjarne) [inline src/AppKit/NSCell.cs:53]:
```suggestion
				frame, startCap.GetHandle (),
				centerFill.GetHandle (),
				endCap.Handle GetHandle (),
```

File: src/AppKit/NSCell.cs [hunk 13]
```diff
@@ -66,15 +68,15 @@ public partial class NSCell {
...
-				bottomRightCorner != null ? bottomRightCorner.Handle : IntPtr.Zero,
+				frame, topLeftCorner is not null ? topLeftCorner.Handle : IntPtr.Zero,
+				topEdgeFill is not null ? topEdgeFill.Handle : IntPtr.Zero,
+				topRightCorner is not null ? topRightCorner.Handle : IntPtr.Zero,
+				leftEdgeFill is not null ? leftEdgeFill.Handle : IntPtr.Zero,
+				centerFill is not null ? centerFill.Handle : IntPtr.Zero,
+				rightEdgeFill is not null ? rightEdgeFill.Handle : IntPtr.Zero,
+				bottomLeftCorner is not null ? bottomLeftCorner.Handle : IntPtr.Zero,
+				bottomEdgeFill is not null ? bottomEdgeFill.Handle : IntPtr.Zero,
+				bottomRightCorner is not null ? bottomRightCorner.Handle : IntPtr.Zero,
```
Comment (rolfbjarne) [inline src/AppKit/NSCell.cs:79]:
```suggestion
				frame, topLeftCorner.GetHandle (),
				topEdgeFill.GetHandle (),
				topRightCorner.GetHandle (),
				leftEdgeFill.GetHandle (),
				centerFill.GetHandle (),
				rightEdgeFill.GetHandle (),
				bottomLeftCorner.GetHandle (),
				bottomEdgeFill.GetHandle (),
				bottomRightCorner.GetHandle (),
```

File: src/AppKit/NSDocument.cs [hunk 14]
```diff
@@ -34,7 +36,7 @@ void SelectorCallback (NSDocument source, bool didDuplicate, IntPtr contextInfo)
 
 		public void DuplicateDocument (DuplicateCallback callback)
```
Comment (rolfbjarne) [inline src/AppKit/NSDocument.cs:37]:
```suggestion
		public void DuplicateDocument (DuplicateCallback? callback)
```

File: src/AppKit/NSFont.cs [hunk 15]
```diff
@@ -25,9 +27,9 @@ public static NSFont FromCTFont (CTFont font)
 #else
 		[Mac (10, 13)]
 #endif
-		public unsafe CGRect [] GetBoundingRects (CGGlyph [] glyphs)
+		public unsafe CGRect [] GetBoundingRects (CGGlyph []? glyphs)
```
Comment (rolfbjarne) [inline src/AppKit/NSFont.cs:30]: `glyphs`  shouldn't be null according to the code.

File: src/AppKit/NSFont.cs [hunk 16]
```diff
@@ -47,9 +49,9 @@ public static NSFont FromCTFont (CTFont font)
 #else
 		[Mac (10, 13)]
 #endif
-		public unsafe CGSize [] GetAdvancements (CGGlyph [] glyphs)
+		public unsafe CGSize [] GetAdvancements (CGGlyph []? glyphs)
```
Comment (rolfbjarne) [inline src/AppKit/NSFont.cs:52]: Same

File: src/AppKit/NSTableView.cs [hunk 17]
```diff
@@ -30,13 +30,15 @@
...
 
 	public partial class NSTableView {
-		public NSTableViewSource Source {
+		public NSTableViewSource? Source {
 			get {
 				var d = WeakDelegate as NSTableViewSource;
-				if (d != null)
+				if (d is not null)
 					return d;
 				return null;
```
Comment (rolfbjarne) [inline src/AppKit/NSTableView.cs:43]: This can be simplified as:

```cs
return WeakDelegate as NSTableViewSource;
```

File: src/AppKit/NSView.cs [hunk 18]
```diff
@@ -31,20 +33,25 @@ sealed class SortData {
...
 			} catch (Exception e) {
-				data.Exception = e;
+				if (data is not null)
+					data.Exception = e;
 				return (nint) (long) NSComparisonResult.Same;
 			}
 		}
 
-		public unsafe void SortSubviews (Func<NSView, NSView, NSComparisonResult> comparer)
+		public unsafe void SortSubviews (Func<NSView, NSView, NSComparisonResult>? comparer)
```
Comment (rolfbjarne) [inline src/AppKit/NSView.cs:52]: `comparer` shouldn't be null according to the code.

However, the two arguments to compare should allow null:

```suggestion
		public unsafe void SortSubviews (Func<NSView?, NSView?, NSComparisonResult> comparer)
```

File: src/AppKit/NSView.cs [hunk 19]
```diff
@@ -20,8 +22,8 @@ public partial class NSView {
 #endif
 
 		sealed class SortData {
-			public Exception Exception;
-			public Func<NSView, NSView, NSComparisonResult> Comparer;
+			public Exception? Exception;
+			public Func<NSView, NSView, NSComparisonResult>? Comparer;
```
Comment (rolfbjarne) [inline src/AppKit/NSView.cs:26]: If you add a constructor, you can avoid making the `Comparer` field nullable.

File: src/AppKit/NSView.cs [hunk 20]
```diff
@@ -31,20 +33,25 @@ sealed class SortData {
 #endif
 		static nint view_compare (IntPtr view1, IntPtr view2, IntPtr context)
 		{
-			var data = (SortData) GCHandle.FromIntPtr (context).Target;
+			var data = GCHandle.FromIntPtr (context).Target as SortData;
```
Comment (rolfbjarne) [inline src/AppKit/NSView.cs:36]: Why was this change necessary? `data` should always be `SortData`, so the old code is safe.

File: src/AppKit/NSView.cs [hunk 21]
```diff
@@ -31,20 +33,25 @@ sealed class SortData {
...
-				var a = (NSView) Runtime.GetNSObject (view1);
-				var b = (NSView) Runtime.GetNSObject (view2);
-				return (nint) (long) data.Comparer (a, b);
+				if (Runtime.GetNSObject (view1) is NSView a
+						&& Runtime.GetNSObject (view2) is NSView b) {
+					var result = data?.Comparer?.Invoke (a, b);
+					return result is null ? (nint) (long) NSComparisonResult.Same : (nint) (long) result;
+				} else {
+					return (nint) (long) NSComparisonResult.Same;
+				}
```
Comment (rolfbjarne) [inline src/AppKit/NSView.cs:44]: You can mostly use the old code if you change the comparison function to allow null items (see further below):

```suggestion
				var a = Runtime.GetNSObject (view1) as NSView;
				var b = Runtime.GetNSObject (view2) as NSView;
				return (nint) (long) data.Comparer (a, b);
```

File: src/AppKit/NSCollectionViewDelegate.cs [hunk 22]
```diff
@@ -18,9 +20,9 @@ public virtual NSDragOperation ValidateDrop (NSCollectionView collectionView, NS
 	public partial class NSCollectionViewDelegateFlowLayout {
 		[Mac (10, 11)]
 		[Obsolete ("Use 'ValidateDropOperation (NSCollectionView collectionView, NSDraggingInfo draggingInfo, ref NSIndexPath proposedDropIndexPath, ref NSCollectionViewDropOperation proposedDropOperation)' instead.")]
-		public virtual NSDragOperation ValidateDrop (NSCollectionView collectionView, NSDraggingInfo draggingInfo, out NSIndexPath proposedDropIndexPath, out NSCollectionViewDropOperation proposedDropOperation)
+		public virtual NSDragOperation? ValidateDrop (NSCollectionView collectionView, NSDraggingInfo draggingInfo, out NSIndexPath proposedDropIndexPath, out NSCollectionViewDropOperation proposedDropOperation)
```
Comment (rolfbjarne) [inline src/AppKit/NSCollectionViewDelegate.cs:23]: This is a breaking change, since NSDragOperation is an enum.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/AppKit/NSCollectionViewDelegate.cs (hunk 22 above)
Comment (mandel-macaque) [reply src/AppKit/NSCollectionViewDelegate.cs:23]: Tha tis a typo. Went too fast.

File: src/AppKit/ActionDispatcher.cs [hunk 23]
```diff
@@ -73,46 +75,46 @@ public ActionDispatcher ()
...
 			ctarget.Activated += handler;
 			return ctarget;
 		}
 
-		public static void RemoveAction (NSObject target, EventHandler handler)
+		public static void RemoveAction (NSObject? target, EventHandler handler)
 		{
-			ActionDispatcher ctarget = target as ActionDispatcher;
-			if (ctarget == null)
+			var ctarget = target as ActionDispatcher;
```
Comment (stephen-hawley) [inline src/AppKit/ActionDispatcher.cs:90]: Why not?
```
if (target is ActionDispatcher ctarget) {
    ctarget.Activated -= handler;
}
```

File: src/AppKit/ActionDispatcher.cs [hunk 24]
```diff
@@ -40,25 +42,25 @@ internal class ActionDispatcher : NSObject
 		const string dkey = "__monomac_internal_ActionDispatcher_doubleActivated:";
 		public static Selector Action = new Selector (skey);
 		public static Selector DoubleAction = new Selector (dkey);
-		public EventHandler Activated;
-		public EventHandler DoubleActivated;
+		public EventHandler? Activated;
```
Comment (stephen-hawley) [inline src/AppKit/ActionDispatcher.cs:45]: In previous places I've worked, we liked to set up `EventHandler` members like this:
```
public EventHandler Activated = (s, e) => { };
```
This lets you turn the `OnActivated` into a one-liner:
```
[Preserve, Export (skey)]
public void OnActivated (NSObject sender) => handler (sender, EventArgs.Empty);
```

**Review by stephen-hawley — CHANGES_REQUESTED** [review]

File: src/AppKit/ActionDispatcher.cs (hunk 24 above)
Comment (rolfbjarne) [reply src/AppKit/ActionDispatcher.cs:45]: That uses more memory, because you'd always create an `EventHandler` instance, even if nobody adds an event handler.

Comment (mandel-macaque) [reply src/AppKit/ActionDispatcher.cs:45]: That was the main reason I did not use that. @stephen-hawley  I did see you using that code, I though about it but considered th emem usage and the fact that it is an ActionDispatcher.

Comment (stephen-hawley) [reply src/AppKit/ActionDispatcher.cs:45]: How big is an empty `EventHandler` instance compared to the extra code for `null` checks?

**Review by rolfbjarne — APPROVED** [review]

File: src/AppKit/NSAccessibility.cs [hunk 25]
```diff
@@ -90,13 +88,13 @@ public static void PostNotification (NSObject element, NSString notification)
...
 		static extern IntPtr NSAccessibilityRoleDescription (IntPtr role, IntPtr subrole);
 
-		public static string GetRoleDescription (NSString role, NSString subrole)
+		public static string? GetRoleDescription (NSString role, NSString? subrole)
 		{
-			if (role == null)
+			if (role is null)
 				throw new ArgumentNullException ("role");
 
 			IntPtr subroleHandle;
```
Comment (mandel-macaque) [inline src/AppKit/NSAccessibility.cs:96]:
```suggestion
			var subroleHandle = subrole.GetHandle ();
```

File: src/AppKit/NSAccessibility.cs [hunk 26]
```diff
@@ -90,13 +88,13 @@ public static void PostNotification (NSObject element, NSString notification)
...
-			if (subrole == null)
+			if (subrole is null)
 				subroleHandle = IntPtr.Zero;
 			else
 				subroleHandle = subrole.Handle;
```
Comment (mandel-macaque) [inline src/AppKit/NSAccessibility.cs:100]:
```suggestion
```
